tests/dialects/test_llvm.py

```diff
@@ -31,16 +32,66 @@ def test_llvm_pointer_ops():
 
     assert isinstance(nullptr.nullptr.typ, llvm.LLVMPointerType)
     assert isinstance(nullptr.nullptr.typ.type, builtin.NoneAttr)
-    assert isinstance(nullptr.nullptr.typ.addr_space, builtin.NoneAttr)
+    assert isinstance(nullptr.nullptr.typ.addr_space, builtin.NoneAttr)    
 
 
 def test_llvm_pointer_type():
     assert llvm.LLVMPointerType.typed(builtin.i64).is_typed()
     assert llvm.LLVMPointerType.typed(builtin.i64).type is builtin.i64
     assert isinstance(
         llvm.LLVMPointerType.typed(builtin.i64).addr_space, builtin.NoneAttr)
-
     assert not llvm.LLVMPointerType.opaque().is_typed()
     assert isinstance(llvm.LLVMPointerType.opaque().type, builtin.NoneAttr)
     assert isinstance(llvm.LLVMPointerType.opaque().addr_space,
                       builtin.NoneAttr)
+    
+    
+def test_llvm_getelementptr_op_invalid_construction():
+    size = arith.Constant.from_int_and_width(1, 32)
+    ptr = llvm.AllocaOp.get(size, builtin.i32)
+    opaque_ptr = llvm.AllocaOp.get(size, builtin.i32, as_untyped_ptr=True)
+
+    # check that passing an opaque pointer to GEP without a pointee type fails
+    with pytest.raises(ValueError):
+        llvm.GEPOp.get(opaque_ptr, llvm.LLVMPointerType.typed(builtin.i32), [1])
+
+    # check that non-pointer arguments fail
+    with pytest.raises(ValueError):
+        llvm.GEPOp.get(size, llvm.LLVMPointerType.opaque())
+
+    # check that non-pointer result types fail
+    with pytest.raises(ValueError):
+        llvm.GEPOp.get(ptr, builtin.i32, [1]) #type: ignore
+
+
+def test_llvm_getelementptr_op():
+    size = arith.Constant.from_int_and_width(1, 32)
+    ptr = llvm.AllocaOp.get(size, builtin.i32)
+    ptr_typ = ptr.res.typ
+    opaque_ptr = llvm.AllocaOp.get(size, builtin.i32, as_untyped_ptr=True)
+
+    # check that construction with static-only offsets and inbounds attr works:
+    gep1 = llvm.GEPOp.get(ptr,ptr_typ, [1], inbounds=True)
+
+    assert 'inbounds' in gep1.attributes
+    assert gep1.result.typ == ptr_typ
+    assert gep1.ptr == ptr.res
+    assert 'elem_type' not in gep1.attributes
+    assert len(gep1.rawConstantIndices.data) == 1
+    assert len(gep1.ssa_indices) == 0
+
+    # check that construction with opaque pointer works:
+    gep2 = llvm.GEPOp.get(opaque_ptr, ptr_typ, [1], pointee_type=builtin.i32)
+
+    assert 'elem_type' in gep2.attributes
+    assert 'inbounds' not in gep2.attributes
+    assert gep2.result.typ == ptr_typ
+    assert len(gep1.rawConstantIndices.data) == 1
+    assert len(gep1.ssa_indices) == 0
+
+    # check GEP with mixed args
+    gep3 = llvm.GEPOp.get(ptr, ptr_typ, [1, -2147483648], [size])
```

Comment:
Should we get this "magic number in a better way?"
print(numpy.iinfo(numpy.int32))
https://thispointer.com/get-maximum-minimum-values-for-ints-in-python/#get-maximum-and-minimum-values-of-int-using-sysmaxsize-in-python3

Reply:
Oh that's really interesting, will implement

Reply:
I can do that yep